helix-term/src/commands.rs

```diff
@@ -1282,16 +1282,23 @@ pub fn scroll(cx: &mut Context, offset: usize, direction: Direction) {
         .max(view.offset.row + scrolloff)
         .min(last_line.saturating_sub(scrolloff));
 
-    let head = pos_at_coords(text, Position::new(line, cursor.col), true); // this func will properly truncate to line end
+    // If cursor needs moving, replace primary selection
+    if line != cursor.row {
+        let head = pos_at_coords(text, Position::new(line, cursor.col), true); // this func will properly truncate to line end
 
-    let anchor = if doc.mode == Mode::Select {
-        range.anchor
-    } else {
-        head
-    };
+        let anchor = if doc.mode == Mode::Select {
+            range.anchor
+        } else {
+            head
+        };
 
-    // TODO: only manipulate main selection
-    doc.set_selection(view.id, Selection::single(anchor, head));
+        // replace primary selection with an empty selection at cursor pos
+        let prim_sel = Range::new(anchor, head);
```

Comment:
Do we need to clone, or is it possible just to mutate the primary range that we already got from above?

Reply:
Selections are currently not mutably accessible outside Document. I believe this protects the Integrity of the selections, e.g. not having overlapping selections (see: https://github.com/helix-editor/helix/blob/aaa42e1a69e4bf88e6761e51085957201295ca52/helix-core/src/selection.rs#L523)

Ranges and Selections are pretty lightweight too: Range is 3 `usize` and Selection is 1 vec of Ranges + 1 `usize`

Reply:
Yeah, but `Range: Copy`, and `Selection::primary` returns an owned Range, which is its own copy, if I'm not mistaken.

Reply:
I'm not sure to get your point. Currently, `Document::set_selection` is the only way to modify selections and it asks for a owned `Selection`, so a copy of Selection is required anyway. The Ranges inside Selections are not mutable in any way so a copy or a new range is also required.

Reply:
Sorry, I see what you mean now: `Selection::replace` takes a `mut self`, so cloning is necessary. Never mind me.